### PR TITLE
executor: convert inner key to the type of outer key’s when build hash table (#8697)

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
HashLeftJoin executor can't join two decimals with different fraction rightly; (#8697)

### What is changed and how it works?
When build inner hash table, convert inner keys to the type of outer keys like what IndexLookUpJoin do;

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8873)

### CODE REVIEW

**Comment by zz-jason** [comment]
@qw4990 Please add some proper labels. For example, component, type.

File: executor/join.go [hunk 1]
```diff
@@ -185,10 +185,26 @@ func (e *HashJoinExec) getJoinKeyFromChkRow(isOuterKey bool, row chunk.Row, keyB
 	}
 
 	keyBuf = keyBuf[:0]
-	keyBuf, err = codec.HashChunkRow(e.ctx.GetSessionVars().StmtCtx, keyBuf, row, allTypes, keyColIdx)
-	if err != nil {
-		err = errors.Trace(err)
+	sc := e.ctx.GetSessionVars().StmtCtx
+	for idx, i := range keyColIdx {
+		d := row.GetDatum(i, allTypes[i])
```
Comment (zz-jason) [inline executor/join.go]: It's inefficient to convert the data inside a `Chunk` to `Datum` then convert the data to the target type. Can we directly convert the data inside the `Chunk` to the target type?

File: executor/join_test.go [hunk 2]
```diff
@@ -988,3 +990,52 @@ func (s *testSuite2) TestHashJoin(c *C) {
...
+	rows := result.Rows()
+	joinRow := fmt.Sprintf("%v", rows[1])
+	isIndexJoin := strings.Contains(joinRow, "IndexJoin")
+	c.Assert(isIndexJoin, Equals, true)
+	//| Projection_5             | 1.00  | root | t1.a                                                                        |
+	//| └─IndexJoin_9            | 1.00  | root | inner join, inner:IndexReader_8, outer key:t1.b, inner key:t2.c             |
+	//|   ├─TableReader_11       | 1.00  | root | data:TableScan_10                                                           |
+	//|   │ └─TableScan_10       | 1.00  | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo                 |
+	//|   └─IndexReader_8        | 0.00  | root | index:IndexScan_7                                                           |
+	//|     └─IndexScan_7        | 0.00  | cop  | table:t2, index:c, range: decided by [t1.b], keep order:false, stats:pseudo |
```
Comment (zz-jason) [inline executor/join_test.go]: This kind of comment is ugly and hard to maintain. We may change the behavior of planner in the future, and then this comment is outdated, but we can not be aware to change this comment.

Maybe we should remove it. Or we can add this test in the integration test, not in this unit test:
```sql
-- first check the execution plan.
explain explain select /*+ TIDB_INLJ(t1,t2) */ t1.a from t t1 join t t2 on t1.b = t2.c;

-- then check the execution result.
select /*+ TIDB_INLJ(t1,t2) */ t1.a from t t1 join t t2 on t1.b = t2.c;
```

**Comment by XuHuaiyu** [comment]
@XuHuaiyu

**Comment by XuHuaiyu** [comment]
@XuHuaiyu

File: executor/join.go [hunk 3]
```diff
@@ -180,16 +180,40 @@ func (e *HashJoinExec) getJoinKeyFromChkRow(isOuterKey bool, row chunk.Row, keyB
...
+			// convert inner key to the type of outer key
+			converted, err := d.ConvertTo(sc, outerType)
+			if err != nil {
+				return false, false, keyBuf, errors.Trace(err)
+			}
+			cmp, err := converted.CompareDatum(sc, &d)
+			if err != nil {
+				return false, false, keyBuf, errors.Trace(err)
+			}
+			if cmp != 0 {
```
Comment (eurekaka) [inline executor/join.go]: Can you give an example where the converted datum is unequal with the original one?

**Comment by zz-jason** [comment]
@qw4990 any update?

**Comment by qw4990** [comment]
This problem(#8697) is solved by this PR(https://github.com/pingcap/tidb/pull/8930);

**Pull request closed without merging**